### ISSUE TITLE
Hey a components ChangeDetectorRef without using its constructor

### ISSUE THREAD

**Issue by tsproesser** [angular/angular#38626]
# 🚀 feature request

### Relevant Package
@angular/core

### Description
**As far as I know** the only possibility to get a components ChangeDetectorRef is by passing it in its constructor and accessing it this way. It can't be get by the Injector since it's some type of "special injection" being passed as a reference on construction (see https://stackoverflow.com/posts/comments/112463363, https://stackoverflow.com/posts/comments/112425354). In some "framework-like" situations doing this can be a relevant scenario (e.g. I've implemented a component base class that should be able to access its current CDR. Right now I do this by implementing an abstract method retrieving the actual CDR - but that's not very neat).

### Describe the solution you'd like
Actually I'm not sure that there is none! But maybe noone out there (Stackoverflow) knows how to do it? Maybe pass an instance to the Injector to be able to get instance-bound dependencies? Or resolve an instance-bound Injector?

### Describe alternatives you've considered
If someone tells me how to access the CDR via private properties I could for now maybe go with a "hack"?

**Comment by Airblader** [comment]
> I've implemented a component base class that should be able to access its current CDR

Normally you inject the CDR using DI by declaring it in the constructor. I don't see a reason why a base class couldn't do the same. Of course no DI will happen there, but any child class extending from it has to call its super constructor and pass the CDR, which it then itself gets through DI.

**Comment by tsproesser** [comment]
> > I've implemented a component base class that should be able to access its current CDR
> 
> Normally you inject the CDR using DI by declaring it in the constructor. I don't see a reason why a base class couldn't do the same. Of course no DI will happen there, but any child class extending from it has to call its super constructor and pass the CDR, which it then itself gets through DI.
> 
> Even if you were to just get it from the injector directly you'd still need access to the injector anyway. You'll always have to pass "something" to the base class.

Yes, that's absolutely right. But since this base class is used in some sort of Framework (actually there is a chain of some base classes) I started injecting only a certain class in all the components, so that I can dynamicly adopt the required services there. But adding CDR to that class surely isn't possible (because of that "special injection thing"). Neither do I want to pass the CDR of almost all the components to the base class. That just seams pretty ugly - and the base class only needs it in a few scenarios (only for some of the derived classes).

**Comment by Airblader** [comment]
As an unrested idea: if you know the selectors for which you need the CDR, create a directive with those selectors which injects the CDR. Then you can get that directive instance from the injector and access the CDR through it. From the usage side of the components this is completely transparent.

**Comment by tsproesser** [comment]
> As an untested idea: if you know the selectors for which you need the CDR, create a directive with those selectors which injects the CDR. Then you can get that directive instance from the injector and access the CDR through it. From the usage side of the components this is completely transparent.

Sounds like a nice workaround! But I have no idea how this selector could look like? Actualy every component could implement this base class. Maybe I just don't get you right? And couldn't this maybe have a noticable performance impact? I really need the CDR's reference in my base component in only a few scenarios.

**Comment by SebastianPodgajny** [comment]
Something like this works for you?

```typescript
import {
  ɵComponentDef as ComponentDef,
  ɵNG_COMP_DEF as NG_COMP_DEF,
  ɵɵdirectiveInject as directiveInject,
} from '@angular/core';

export const DirectiveInject = (type: any): PropertyDecorator => {
  return function(target, propertyKey) {
    const isJIT: boolean = !target.constructor[NG_COMP_DEF];
    const inject = () => {
      const compDef: ComponentDef<any> = target.constructor[NG_COMP_DEF];
      const compFactory = compDef.factory || target.constructor['ɵfac'];
      compDef.factory = function(...args: any[]) {
        const instance = compFactory(...args);
        instance[propertyKey] = directiveInject(type);
        return instance;
      }
    }

    isJIT ? Promise.resolve().then(inject) : inject();
  }
}

export class AppComponent {
  @DirectiveInject(ChangeDetectorRef)
  public cdRef: ChangeDetectorRef;
}
```

**Comment by tsproesser** [comment]
> Something like this works for you?
> 
> ```ts
> import {
>   ɵComponentDef as ComponentDef,
>   ɵNG_COMP_DEF as NG_COMP_DEF,
>   ɵɵdirectiveInject as directiveInject,
> } from '@angular/core';
> 
> export const DirectiveInject = (type: any): PropertyDecorator => {
>   return function(target, propertyKey) {
>     const isJIT: boolean = !target.constructor[NG_COMP_DEF];
>     const inject = () => {
>       const compDef: ComponentDef<any> = target.constructor[NG_COMP_DEF];
>       const compFactory = compDef.factory || target.constructor['ɵfac'];
>       compDef.factory = function(...args: any[]) {
>         const instance = compFactory(...args);
>         instance[propertyKey] = directiveInject(type);
>         return instance;
>       }
>     }
> 
>     isJIT ? Promise.resolve().then(inject) : inject();
>   }
> }
> 
> export class AppComponent {
>   @DirectiveInject(ChangeDetectorRef)
>   public cdRef: ChangeDetectorRef;
> }
> ```

Throws an error:

zone.js:1064 Unhandled Promise rejection: Cannot read property 'factory' of undefined.

**Comment by SebastianPodgajny** [comment]
Do you use Ivy?

**Comment by tsproesser** [comment]
Yes, latest angular version (10.0.14) with Ivy enabled.

**Comment by SebastianPodgajny** [comment]
https://stackblitz.com/edit/angular-ivy-nagp3n

**Comment by tsproesser** [comment]
Ah, ok - sorry. It actually worked but just dropped this Error in the console. Anyway: when using it in a BaseClass it doesn't:

https://stackblitz.com/edit/angular-ivy-4edibh?file=src%2Fapp%2Fapp.component.html

**Comment by Totati** [comment]
> Ah, ok - sorry. It actually worked but just dropped this Error in the console. Anyway: when using it in a BaseClass it doesn't:
> 
> https://stackblitz.com/edit/angular-ivy-4edibh?file=src%2Fapp%2Fapp.component.html

That's because you're not using `@Component` or `@Directive` decorator on your base. [Here's](https://stackblitz.com/edit/angular-ivy-ydprmt?file=src%2Fapp%2Fapp.component.ts) a working example.

**Comment by tsproesser** [comment]
> > Ah, ok - sorry. It actually worked but just dropped this Error in the console. Anyway: when using it in a BaseClass it doesn't:
> > https://stackblitz.com/edit/angular-ivy-4edibh?file=src%2Fapp%2Fapp.component.html
> 
> That's because you're not using `@Component` or `@Directive` decorator on your base. [Here's](https://stackblitz.com/edit/angular-ivy-ydprmt?file=src%2Fapp%2Fapp.component.ts) a working example.

Sorry, was just too lazy. You're surely right, but as you can see, the base CdRef is still null: "CdRefBase exists: false"

**Comment by tsproesser** [comment]
Any news on this? Been a while and I still don't have any clue :(

**Comment by angelaki** [comment]
Just in case you're interested: my use-case is an abstract base component that needs the `cdr` in some cases. Putting it in the constructor needs all components inheriting from it to pass it to it's base. That's why I'd like to grab it somehow just by knowing _myself_ (`getCdr(this)`).

But yeah, it's a pretty special use-case.

**Comment by Airblader** [comment]
In that case you're probably interested in https://github.com/angular/angular/issues/34425#issuecomment-1117473457 where the team just stated that it is on the roadmap.
